core/headerchain.go

```diff
@@ -165,7 +165,9 @@ func (hc *HeaderChain) WriteHeader(header *types.Header) (status WriteStatus, er
 	// If the total difficulty is higher than our known, add it to the canonical chain
 	// Second clause in the if statement reduces the vulnerability to selfish mining.
 	// Please refer to http://www.cs.cornell.edu/~ie53/publications/btcProcFC.pdf
-	if externTd.Cmp(localTd) > 0 || (externTd.Cmp(localTd) == 0 && mrand.Float64() < 0.5) {
+	if externTd.Cmp(localTd) > 0 ||
+		(externTd.Cmp(localTd) == 0 && header.Number.Uint64() < hc.CurrentHeader().Number.Uint64()) ||
```

Comment:
We already retrieve the current header in line 152, perhaps lets stash that into a variable and reuse here. That way it's safer race wise since we only retrieve it once and keep operating on the same data.